### PR TITLE
[Bugfix] Metis postings render markdown correctly

### PR DESCRIPTION
**Pull request by lschlesinger**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).

### Motivation and Context

- Solves #3576
- Create a dedicated `postings-markdown-editor` to avoid code duplication
- Adapts the default commands (shown formatting options) for the postings editor, which are not limited to the most important ones

### Description

The array of allowed html tags and attributes in `answer-post.component.ts` and `post.component.ts` was too restrictive. Almost every formatting feature the markdown editor was offering got lost. 

By extending `allowedHtmlTags` and `allowedHtmlAttributes`, the browser is able to render, what is promised by the markdown preview without introducing possible XSS vulnerabilities. 

Headings ("#h1", "##h2") rely on the attribute `id` to be allowed (which we do not want), and can therefore not be rendered correctly.

The "format-short-cut"-bar of the markdown editor was reduced to formats that are rendered correctly and most probably useful. (Un-)Ordered lists are also rendered correctly but are, for the sake of simplicity, excluded from the "format-short-cut"-bar. Simply using "-" (for unordered lists) and "1." (for ordered lists) should be intuitive.

Bildschirmfoto 2021-07-07 um 16 59 21

### Steps for Testing

1. Log in to Artemis
2. Navigate to a course with postings enabled and within that course to an exercise or lecture page -> see discussion section on the right
3. Create some posts and answer posts using the markdown editor features, e.g. such as shown in the input screenshot etc.
4. Check if the (answer) post is displayed as shown in the preview after saving it.

### Screenshots

**Input**:
Bildschirmfoto 2021-07-07 um 17 02 35

**Before**:
Bildschirmfoto 2021-07-07 um 17 03 11

**After**:
Bildschirmfoto 2021-07-07 um 17 03 40

### CODE REVIEW

File: src/main/webapp/app/overview/postings/answer-post/answer-post.component.ts [hunk 1]
```diff
@@ -31,9 +31,33 @@ export class AnswerPostComponent implements OnInit {
     EditorMode = EditorMode;
     courseId: number;
 
-    // Only allow certain html tags and no attributes
-    allowedHtmlTags: string[] = ['a', 'b', 'strong', 'i', 'em', 'mark', 'small', 'del', 'ins', 'sub', 'sup', 'p'];
```
Comment (lschlesinger) [inline src/main/webapp/app/overview/postings/answer-post/answer-post.component.ts:35]: @krusche , is there any reason why we where so restrictive with html tags. This made most of the markdown being lost while converting into html. IMO the added tags and attributes are not critical w.r.t. security... am I wrong?

**Review by willergeorg — CHANGES_REQUESTED** [review]
Tested on TS2. There were some deviations between the preview of the actual display for me (see screenshots below).

The used text:
image3

The preview:
image

The final posting:
image2

**Review by NicolasUllmann — CHANGES_REQUESTED** [review]
Tested on TS2. I ran into the same problem as Georg mentioned above.

**Review by real-dmy — CHANGES_REQUESTED** [review]
Tested on TS1. Everything works fin except the issue mentioned by Geord above.

**Comment by lschlesinger** [comment]
@NicolasUllmann @willergeorg @real-dmy thanks for testing and pointing out the remaining issue. h2 and h3 are now also rendered correctly. I removed the formula option, as it was only available for post creation which might be confusing anyways.

**Comment by lschlesinger** [comment]
After talking to @krusche, I have to investigate possible security issues first before allowing the added html tags and attributes. -> PR on hold.

**Comment by lschlesinger** [comment]
@FrankeLukas I am aware of the issue #2585 and spent some time trying to succeed with a XSS attack using [suggested payload ](https://github.com/cujanovic/Markdown-XSS-Payloads/blob/master/Markdown-XSS-Payloads.txt#L40) - but I didn't. I saw, that you created the PR which restricted the allowed html tags and attributes back then and I therefore assume, that you are familiar with the topic to some degree. As far as I can tell, the additional allowed tags (introduced with this PR), do not imply less security with regard to XSS attacks. 

I would like to at least offer the most important functionality for creating posts in the programming context. This is also aligned with the creator of the issue #3576 (which caught my attention in the first place) to whom I have spoken. These would be code, codeblocks, references (these commands would be displayed in the markdown editor - others are deleted to not suggest anything will be rendered). Also I think lists are helpful to provide some structure to posts with larger content.

File: src/main/webapp/app/overview/postings/answer-post/answer-post.component.ts [hunk 2]
```diff
@@ -26,23 +26,23 @@ export class AnswerPostComponent implements OnInit {
...
     isLoading = false;
     isEditMode: boolean;
     EditorMode = EditorMode;
     courseId: number;
 
-    // Only allow certain html tags and no attributes
-    allowedHtmlTags: string[] = ['a', 'b', 'strong', 'i', 'em', 'mark', 'small', 'del', 'ins', 'sub', 'sup', 'p'];
-    allowedHtmlAttributes: string[] = ['href'];
+    // Only allow certain html tags and attributes
+    allowedHtmlTags: string[] = ['a', 'b', 'strong', 'i', 'em', 'mark', 'small', 'del', 'ins', 'sub', 'sup', 'p', 'ins', 'blockquote', 'pre', 'code', 'span', 'li', 'ul'];
```
Comment (MaisiKoleni) [inline src/main/webapp/app/overview/postings/answer-post/answer-post.component.ts:36]: `ins` was already in the list. I am not an HTML expert, but the tags look safe to me.
```suggestion
    allowedHtmlTags: string[] = ['a', 'b', 'strong', 'i', 'em', 'mark', 'small', 'del', 'ins', 'sub', 'sup', 'p', 'blockquote', 'pre', 'code', 'span', 'li', 'ul'];
```
We could maybe even allow more tags that are similar, like `ol` for ordered lists.

Comment (lschlesinger) [reply src/main/webapp/app/overview/postings/answer-post/answer-post.component.ts]: 9cc3183

**Review by willergeorg — APPROVED** [review]
Tested on TS1. Seems to work good for me now.

**Review by ArtemTsurikov — DISMISSED** [review]
Tested on TS1, looks good to me!

**Review by NicolasUllmann — APPROVED** [review]
Tested on TS1. Everything worked as expected.

**Comment by lschlesinger** [comment]
For security reasons, I had to remove `class` from the list of allowed attributes. Fortunately, this does not affect the formatting of the selected options too heavyly. Only the code block formatting is not as advanced as before, i.e. there is  not language specific code highlighting and such things. As a result, the preview differs slightly from the actually rendered saved post:
preview:
Bildschirmfoto 2021-07-09 um 15 02 36
rendered result:
Bildschirmfoto 2021-07-09 um 15 02 44
I guess this is a price we can pay 😄

**Review by Olimbia23 — APPROVED** [review]
Checked the code and I approve @lschlesinger 👍

**Review by real-dmy — APPROVED** [review]
Tested on TS1, works as expected!

**Review by argertboja — APPROVED** [review]
Works as expected on TS2!

**Review by canyelten — APPROVED** [review]
Tested on TS1, works as expected.

**Pull request merged**